pkg/proxy/endpointslicecache.go

```diff
@@ -0,0 +1,190 @@
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"sort"
+
+	discovery "k8s.io/api/discovery/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	utilproxy "k8s.io/kubernetes/pkg/proxy/util"
+	utilnet "k8s.io/utils/net"
+)
+
+// EndpointSliceCache is used as a cache of EndpointSlice information
+type EndpointSliceCache struct {
+	slicesByService  map[types.NamespacedName]map[string]EndpointSliceInfo
+	makeEndpointInfo makeEndpointFunc
+	hostname         string
+	isIPv6Mode       *bool
+	recorder         record.EventRecorder
+}
+
+// EndpointSliceInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+type EndpointSliceInfo struct {
+	Ports     []discovery.EndpointPort
+	Endpoints []EndpointInfo
+}
+
+// EndpointInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+type EndpointInfo struct {
+	Addresses []string
+	Topology  map[string]string
+}
+
+// NewEndpointSliceCache initializes an EndpointSliceCache
+func NewEndpointSliceCache(hostname string, isIPv6Mode *bool, recorder record.EventRecorder, makeEndpointInfo makeEndpointFunc) *EndpointSliceCache {
+	return &EndpointSliceCache{
+		slicesByService:  map[types.NamespacedName]map[string]EndpointSliceInfo{},
+		hostname:         hostname,
+		isIPv6Mode:       isIPv6Mode,
+		makeEndpointInfo: makeEndpointInfo,
+		recorder:         recorder,
+	}
+}
+
+// Update a slice in the cache
+func (e *EndpointSliceCache) Update(endpointSlice *discovery.EndpointSlice) {
+	serviceKey, sliceKey := EndpointSliceCacheKeys(endpointSlice)
+	esInfo := EndpointSliceInfo{
+		Ports:     endpointSlice.Ports,
+		Endpoints: []EndpointInfo{},
+	}
+	for _, endpoint := range endpointSlice.Endpoints {
+		if endpoint.Conditions.Ready == nil || *endpoint.Conditions.Ready == true {
+			esInfo.Endpoints = append(esInfo.Endpoints, EndpointInfo{
+				Addresses: endpoint.Addresses,
+				Topology:  endpoint.Topology,
+			})
+		}
+	}
+	if _, exists := e.slicesByService[serviceKey]; !exists {
+		e.slicesByService[serviceKey] = map[string]EndpointSliceInfo{}
+	}
+	e.slicesByService[serviceKey][sliceKey] = esInfo
+}
+
+// Delete a slice from the cache
+func (e *EndpointSliceCache) Delete(endpointSlice *discovery.EndpointSlice) {
+	serviceKey, sliceKey := EndpointSliceCacheKeys(endpointSlice)
+	delete(e.slicesByService[serviceKey], sliceKey)
+}
+
+// GetEndpointsMap computes an EndpointsMap for a given service
+func (e *EndpointSliceCache) GetEndpointsMap(serviceNN types.NamespacedName) EndpointsMap {
+	endpointsMap := EndpointsMap{}
+	endpointInfoByServicePort := map[ServicePortName]map[string]Endpoint{}
+	endpointSlices := e.slicesByService[serviceNN]
+
+	// We need to build a map of portname -> all ip:ports for that portname.
+	for _, endpointSlice := range endpointSlices {
+		for i := range endpointSlice.Ports {
```

Comment:
if you don't use the index, it's preferable to just have `for _, port := range ...` to avoid having an extra iteration variable floating around.